src/reconcile/rec_write.c

```diff
@@ -1974,40 +1992,79 @@ __rec_split_bnd_grow(WT_SESSION_IMPL *session, WT_RECONCILE *r)
-uint32_t
-__wt_split_page_size(WT_BTREE *btree, uint32_t maxpagesize)
-{
+static uint32_t
+__rec_split_page_size_from_pct(
+    int split_pct, uint32_t maxpagesize, uint32_t allocsize) {
 	uintmax_t a;
 	uint32_t split_size;
 
 	/*
 	 * Ideally, the split page size is some percentage of the maximum page
-	 * size rounded to an allocation unit (round to an allocation unit so
-	 * we don't waste space when we write).
+	 * size rounded to an allocation unit (round to an allocation unit so we
+	 * don't waste space when we write).
 	 */
 	a = maxpagesize;			/* Don't overflow. */
 	split_size = (uint32_t)WT_ALIGN_NEAREST(
-	    (a * (u_int)btree->split_pct) / 100, btree->allocsize);
+	    (a * (u_int)split_pct) / 100, allocsize);
 
 	/*
-	 * Respect the configured split percentage if the calculated split
-	 * size is either zero or a full page. The user has either configured
-	 * an allocation size that matches the page size, or a split
-	 * percentage that is close to zero or one hundred. Rounding is going
-	 * to provide a worse outcome than having a split point that doesn't
-	 * fall on an allocation size boundary in those cases.
+	 * Respect the configured split percentage if the calculated split size
+	 * is either zero or a full page. The user has either configured an
+	 * allocation size that matches the page size, or a split percentage
+	 * that is close to zero or one hundred. Rounding is going to provide a
+	 * worse outcome than having a split point that doesn't fall on an
+	 * allocation size boundary in those cases.
 	 */
 	if (split_size == 0 || split_size == maxpagesize)
-		split_size = (uint32_t)((a * (u_int)btree->split_pct) / 100);
+		split_size = (uint32_t)((a * (u_int)split_pct) / 100);
 
 	return (split_size);
 }
 
 /*
+ * __wt_split_page_size --
+ *	Split page size calculation: we don't want to repeatedly split every
+ * time a new entry is added, so we split to a smaller-than-maximum page size.
+ */
+uint32_t
+__wt_split_page_size(WT_SESSION_IMPL *session, uint32_t maxpagesize)
+{
+	WT_BTREE *btree;
+
+	btree = S2BT(session);
+#define	WT_SPLIT_PCT 90
+
+	return (__rec_split_page_size_from_pct(
+	    WT_SPLIT_PCT, maxpagesize, btree->allocsize));
+}
+
+/*
+ * __wt_alt_split_page_size --
+ *	Alternate boundary split page size calculation: We track an alternate
+ *	boundary we could have split at instead of splitting at the split page
+ *	size. At any given time, we maintain two pages (two split chunks) in the
+ *	memory to be written. If the last split page turns out to be too small,
+ *	we go back into the penultimate chunk and split at this alternate
+ *	boundary. This moves some data from the penultimate chunk to the last
+ *	chunk, hence increasing the size of the last page without decreasing the
+ *	penultimate page size beyond the alternate split size.
+ */
+static uint32_t
+__wt_alt_split_page_size(WT_SESSION_IMPL *session, uint32_t maxpagesize)
+{
+	WT_BTREE *btree;
+
+	btree = S2BT(session);
+#define	WT_ALT_SPLIT_PCT 45
```

Comment:
As discussed offline, this can be anything that is less than the split pct chosen. Since we are restricting split percentage between 50 and 100, alt_split_pct can be anything < 50. 45 is just a random value that I picked for now.